pkg/secretpath/path.go

```diff
@@ -0,0 +1,226 @@
+func Join(elements ...string) string {
+	result := ""
+	for _, e := range elements {
+		e = strings.Trim(e, elemSep)
+		if len(e) != 0 && len(result) == 0 {
+			result += e
+		} else if len(e) != 0 {
+			result += elemSep + e
+		}
+	}
+
+	return Clean(result)
+}
+
+// HasVersion returns true when a version suffix is specified in the path.
+func HasVersion(path string) bool {
+	return hasVersion(path)
+}
+
+// hasVersion returns true when a version suffix is specified in the path.
+func hasVersion(path string) bool {
+	i := strings.LastIndex(path, versionSep)
+	return i >= 0 && i < len(path)-1
+}
+
+// Version returns the version number suffix of a path, returning -1
+// when :latest or negative version numbers are given. If no version
+// suffix is set or it cannot be parsed to an integer, 0 is returned.
+func Version(path string) int {
+	if !hasVersion(path) {
+		return 0
+	}
+
+	if strings.HasSuffix(path, latestSuffix) {
+		return -1
+	}
+
+	i := strings.LastIndex(path, versionSep)
+	version, err := strconv.Atoi(path[i+1:])
+	if err != nil {
+		return 0
+	}
+
+	if version < 0 {
+		return -1
+	}
+
+	return version
+}
+
+// AddVersion adds a version suffix to a given path, removing trailing
+// separators if necessary. If the path already contains a version suffix,
+// it is replaced by the given version number. Negative version numbers
+// are converted to ":latest".
+func AddVersion(path string, version int) string {
+	path = strings.Trim(path, elemSep)
+
+	suffix := latestSuffix
+	if version > 0 {
+		suffix = versionSep + strconv.Itoa(version)
+	}
+
+	i := strings.LastIndex(path, versionSep)
+	if i < 0 {
+		return path + suffix
+	}
+
+	return Clean(path[:i] + suffix)
+}
+
+// Base returns the last element of a path. Trailing separators and version
+// numbers are removed.
+func Base(path string) string {
+	path = Clean(path)
+
+	for i := len(path) - 1; i >= 0; i-- {
+		if path[i] == elemSepByte {
+			return path[i+1:]
+		}
+
+		if path[i] == versionSepByte {
+			path = path[:i]
+		}
+	}
+	return path
+}
+
+// Parent returns all but the last element of a path, removing trailing separators.
+// If a path contains only one element, it returns an empty string.
+func Parent(path string) string {
+	path = Clean(path)
+
+	for i := len(path) - 1; i >= 0; i-- {
+		if path[i] == elemSepByte {
+			return path[:i]
+		}
+	}
+
+	return ""
+}
+
+// Repo returns the first two elements of a path, removing trailing separators.
+// If a path contains less than two elements, it returns the empty string.
+func Repo(path string) string {
+	path = Clean(path)
+
+	count := 0
+	for i := 0; i < len(path); i++ {
+		if path[i] == elemSepByte {
+			count++
+		}
+
+		if count == 2 {
+			return path[:i]
+		}
+	}
+
+	if count == 1 {
+		return path
+	}
+	return ""
+}
+
+// Namespace returns the first element of a path, removing trailing separators.
+func Namespace(path string) string {
+	path = Clean(path)
+	if len(path) == 0 {
```

Comment:
This check is redundant; For the empty string, the loop will have 0 iterations and the final return will be hit immediately.